macros/src/lib.rs

```diff
@@ -34,7 +35,7 @@ impl DerivedTS {
             .params
             .iter()
             .filter(|param| matches!(param, GenericParam::Type(_)))
-            .map(|_| quote! { () });
+            .map(|_| self.unit_type.as_ref().map(|x| x.to_token_stream()).unwrap_or(quote!{ () }));
```

Comment:
The type given to `#[ts(unit_type = "...")]` will be used for all generics, same as `()`, so it must implement all trait bounds from all of the generic parameters

Reply:
Alright, makes sense! Just an idea here: What if the attribute didnt only parse one type, but a list of types? That should just work, right?

Reply:
> What if the attribute didnt only parse one type, but a list of types?

What do you mean? One type per generic?

Reply:
Yeah!

Reply:
I don't think that's necessary though, the type passed in will never be instantiated anywhere in the program, its only purpose is to implement traits and return "null" when `inline()` is called

Reply:
Ah, you're absolutely right. Had to take a 2nd look at the PR, and I just fundamentally misunderstood it. Struggeling so hard with that part of the codebase.

I guess I was thinking the following: What if, instead of matching against `"null"` [here](https://github.com/Aleph-Alpha/ts-rs/pull/217/files#diff-30ff1d0665a8474808fa70cb7d5471b77d797d2a4cfcb2e29c41ff6157bb6bbaR63-R73), we had something like this:
```rust
return quote![
  if TypeId::of::<#generic_ident> == TypeId::of::<#unit_type> {
      #generic_ident_str.to_owned()
  } else {
    <#generic_ident>::inline()
  }
];
```
Then, we'd get rid of the hardcoded "null", and a user could specify any "default generic type", without the need to declare his own "unit" type.
```rs
#[derive(TS)]
#[ts(default_generics = "String, Box<dyn Error>")]
struct Something<S: AsRef<str>, E: Error>();
```

But: `TypeId::of` requires `'static`, and idk if it's even possible to work around that in this case.

Reply:
> Then, we'd get rid of the hardcoded "null", and a user could specify any "default generic type", without the need to declare his own "unit" type.
>
> ```rs
> #[derive(TS)]
> #[ts(default_generics = "String, Box<dyn Error>")]
> struct Something<S: AsRef<str>, E: Error>();
> ```
>But: TypeId::of requires 'static, and idk if it's even possible to work around that in this case.

Yeah, I tried that when I was making the first PR about inlining generics and couldn't get past this, but the bigger problem is, if the user actually wanted at some point to use the struct with the default generic (`String` in this example), inlining the generic would fail, emiiting `S` instead of `String`

Reply:
What we could do instead to get rid of the null is:
- Create an internal type that works like `()` but `inline`s to some constant that is not valid TS
- Replace `()` with said new type
- Compare against this constant instead of `"null"`
- Create a macro that produces an `impl TS` with `ìnline` returning that constant
- Have the user still create a "unit" type for the struct, but use the new macro instead of `#[derive(TS)]`
- Implement the required traits for the type

Reply:
> if the user actually wanted at some point to use the struct with the default generic (String in this example), inlining the generic would fail, emiiting S instead of String

What I mean here is:
```rs
#[derive(TS)]
#[ts(default_generics = "String")]
pub struct Generic<T: ToString> {
    pub x: T,
}

#[derive(TS)]
pub struct Foo {
    #[ts(inline)]
    pub a: Generic<String>, // This will fail, will emit `T` instead of `string` because String == String

   #[ts(inline)]
    pub b: Generic<i32>, // This will not fail because i32 != String

   #[ts(inline)]
    pub b: Generic<&'static str>, // This will not fail because &str != String
}
```

Reply:
Another alternative is, when deciding whether to use `T` or `string`, we need to know if the generic passed to the field is the identifier of one of the struct definition generics or an actual type.
Don't know how we could get this information though

Reply:
> **EDIT**: I think I made it work. Only hacked something together, but it does pass the tests. It's not as robust as the `TypeId` solution before, but it might be preferable to the `"null"` solution - i dont know yet. Here's how it looks like. What do you think?
> 
> ```rust
> return quote!(
>     if std::any::type_name::<#generic_ident>() == std::any::type_name::<#unit_type>() {
>         #generic_ident_str.to_owned()
>     } else {
>         <#generic_ident>::inline()
>     }
> );
> ```
> 
> If that's something you'd like to look in further, I'd be happy to clean my hacky solution up a bit & push it to a branch for you to check out.

This looks interesting, let's take a look!